File: 4037-lexicographically-smallest-palindromic-permutation-greater-than-target/lexicographically-smallest-palindromic-permutation-greater-than-target.py
class Solution:
    def lexPalindromicPermutation(self, s: str, target: str) -> str:
        n = len(s)
        count = [0] * 26
        for char in s:
            count[ord(char) - ord('a')] += 1

        odd_count = 0
        mid_char = -1
        for i in range(26):
            if count[i] % 2 != 0:
                odd_count += 1
                mid_char = i

        if odd_count > 1:
            return ""

        half_len = n // 2
        half_count = [count[i] // 2 for i in range(26)]

        for i in range(half_len, -1, -1):
            current_half_count = half_count.copy()
            valid_prefix = True
            pref = [''] * half_len

            for k in range(i):
                c = ord(target[k]) - ord('a')
                if current_half_count[c] > 0:
                    current_half_count[c] -= 1
                    pref[k] = target[k]
                else:
                    valid_prefix = False
                    break

            if not valid_prefix:
                continue

            if i == half_len:
                if n % 2 != 0:
                    t_mid = target[half_len]
                    if mid_char >= 0 and chr(mid_char + ord('a')) >= t_mid:
                        candidate = self._construct_palindrome(pref, chr(mid_char + ord('a')), current_half_count, n)
                        if candidate > target:
                            return candidate
                else:
                    candidate = self._construct_palindrome(pref, '', current_half_count, n)
                    if candidate > target:
                        return candidate
                continue

            t_char = target[i]
            for next_char in range(ord(t_char) - ord('a') + 1, 26):
                if current_half_count[next_char] > 0:
                    next_half_count = current_half_count.copy()
                    next_half_count[next_char] -= 1
                    pref[i] = chr(next_char + ord('a'))

                    pos = i + 1
                    for c in range(26):
                        while next_half_count[c] > 0:
                            pref[pos] = chr(c + ord('a'))
                            pos += 1
                            next_half_count[c] -= 1

                    mid = chr(mid_char + ord('a')) if n % 2 != 0 else ''
                    candidate = self._build_full_palindrome(pref, mid, n)
                    if candidate > target:
                        return candidate

        return ""

    def _construct_palindrome(self, pref: list, mid: str, remaining_half_count: list, n: int) -> str:
        full_pref = pref.copy()
        pos = len(pref) - remaining_half_count.count(0) # place at end
        pos = len(pref) - sum(remaining_half_count)
        for c in range(26):
            while remaining_half_count[c] > 0:
                full_pref[pos] = chr(c + ord('a'))
                pos += 1
                remaining_half_count[c] -= 1
        return self._build_full_palindrome(full_pref, mid, n)

    def _build_full_palindrome(self, half: list, mid: str, n: int) -> str:
        half_str = "".join(half)
        return half_str + mid + half_str[::-1]

File: 4037-lexicographically-smallest-palindromic-permutation-greater-than-target/lexicographically-smallest-palindromic-permutation-greater-than-target.java
import java.util.*;

class Solution {
    public String lexPalindromicPermutation(String s, String target) {
        int n = s.length();
        int[] count = new int[26];
        for (char c : s.toCharArray()) {
            count[c - 'a']++;
        }

        int oddCount = 0;
        int midChar = -1;
        for (int i = 0; i < 26; i++) {
            if (count[i] % 2 != 0) {
                oddCount++;
                midChar = i;
            }
        }

        if (oddCount > 1) {
            return "";
        }

        int halfLen = n / 2;
        int[] halfCount = new int[26];
        for (int i = 0; i < 26; i++) {
            halfCount[i] = count[i] / 2;
        }

        for (int i = halfLen; i >= 0; i--) {
            int[] currentHalfCount = halfCount.clone();
            boolean validPrefix = true;
            char[] pref = new char[halfLen];

            for (int k = 0; k < i; k++) {
                int c = target.charAt(k) - 'a';
                if (currentHalfCount[c] > 0) {
                    currentHalfCount[c]--;
                    pref[k] = target.charAt(k);
                } else {
                    validPrefix = false;
                    break;
                }
            }

            if (!validPrefix) continue;

            if (i == halfLen) {
                int currentMid = midChar;
                if (n % 2 != 0) {
                    char tMid = target.charAt(halfLen);
                    if (currentMid >= 0 && (currentMid + 'a') > tMid) {
                        String candidate = constructPalindrome(pref, (char) (currentMid + 'a'), currentHalfCount, n);
                        if (candidate.compareTo(target) > 0) return candidate;
                    } else if (currentMid >= 0 && (currentMid + 'a') == tMid) {
                        String candidate = constructPalindrome(pref, (char) (currentMid + 'a'), currentHalfCount, n);
                        if (candidate.compareTo(target) > 0) return candidate;
                    }
                } else {
                    String candidate = constructPalindrome(pref, ' ', currentHalfCount, n);
                    if (candidate.compareTo(target) > 0) return candidate;
                }
                continue;
            }

            char tChar = target.charAt(i);
            for (int nextChar = tChar - 'a' + 1; nextChar < 26; nextChar++) {
                if (currentHalfCount[nextChar] > 0) {
                    int[] nextHalfCount = currentHalfCount.clone();
                    nextHalfCount[nextChar]--;
                    pref[i] = (char) (nextChar + 'a');

                    int pos = i + 1;
                    for (int c = 0; c < 26; c++) {
                        while (nextHalfCount[c] > 0) {
                            pref[pos++] = (char) (c + 'a');
                            nextHalfCount[c]--;
                        }
                    }

                    char mid = n % 2 != 0 ? (char) (midChar + 'a') : ' ';
                    String candidate = buildFullPalindrome(pref, mid, n);
                    if (candidate.compareTo(target) > 0) {
                        return candidate;
                    }
                }
            }
        }

        return "";
    }

    private String constructPalindrome(char[] pref, char mid, int[] remainingHalfCount, int n) {
        char[] fullPref = pref.clone();
        int pos = pref.length;
        for (int c = 0; c < 26; c++) {
            while (remainingHalfCount[c] > 0) {
                fullPref[pos++] = (char) (c + 'a');
                remainingHalfCount[c]--;
            }
        }
        return buildFullPalindrome(fullPref, mid, n);
    }

    private String buildFullPalindrome(char[] half, char mid, int n) {
        StringBuilder sb = new StringBuilder();
        sb.append(half);
        if (n % 2 != 0) {
            sb.append(mid);
        }
        for (int i = half.length - 1; i >= 0; i--) {
            sb.append(half[i]);
        }
        return sb.toString();
    }
}